extensions/pkg/terraformer/state.go

```diff
@@ -91,13 +98,38 @@ func (t *terraformer) GetStateOutputVariables(ctx context.Context, variables ...
 	return output, nil
 }
 
-// IsStateEmpty returns true if the Terraform state is empty, and false otherwise.
+// IsStateEmpty returns true if the Terraform state is empty and the terraformer finalizer
+// is not present on any of the used configmpas and secrets. Otherwise, it returns false.
 func (t *terraformer) IsStateEmpty(ctx context.Context) bool {
+	var needToFinalize = false
+
+	for _, obj := range []client.Object{
+		&corev1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: t.namespace, Name: t.variablesName}},
+		&corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Namespace: t.namespace, Name: t.configName}},
+		&corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Namespace: t.namespace, Name: t.stateName}},
+	} {
+		resourceName := obj.GetName()
+		if err := t.client.Get(ctx, kutil.Key(t.namespace, resourceName), obj); client.IgnoreNotFound(err) != nil {
+			t.logger.Error(err, "failed to get resource", "name", resourceName)
+			return false
+		}
+
+		resourceHasFinalizer, err := extensionscontroller.HasFinalizer(obj, TerraformerFinalizer)
+		if err != nil {
+			t.logger.Error(err, "failed to check finalizer presence for resource", "name", resourceName)
+			return false
+		}
+		if resourceHasFinalizer {
+			needToFinalize = true
+			break
```

Comment:
What about directly return `false` here and skipping the rest?

Reply:
Done.